### ISSUE TITLE
Setting CUDA function attributes

### ISSUE THREAD

**Issue by andravin** [cupy/cupy#2118]
RawKernel is very useful, but currently there appears to be no interface to set the CUDA function attributes.

This is necessary in order to use more than 48KB of shared memory per block in newer GPUs, or to specify the ratio of shared memory to L1 cache (ie, the shared memory "carveout").

Overview of shared memory configuration:
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#shared-memory-7-x

**cudaFuncSetAttribute** documentation:
https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__HIGHLEVEL.html#group__CUDART__HIGHLEVEL_1g422642bfa0c035a590e4c43ff7c11f8d

**Comment by leofang** [comment]
If the attributes are settable, then they can also be read via PR #1874 (not merged yet). In fact I think a PR for this feature request is possible based on #1874.

**Comment by grlee77** [comment]
Yeah, this would be a good followup to #1874.

It looks like there are currently only two attributes that can be assigned:
`cudaFuncAttributeMaxDynamicSharedMemorySize`
`cudaFuncAttributePreferredSharedMemoryCarveout`

The range for these needs to be constrained to `cudaDevAttrMaxSharedMemoryPerBlockOptin` and `cudaDevAttrMaxSharedMemoryPerMultiprocessor`, respectively. In current master, it is now possible to query these device attributes via code such as:

```Python
import cupy
d = cupy.cuda.device.Device()
d.attributes['MaxSharedMemoryPerBlockOptin']
d.attributes['MaxSharedMemoryPerMultiprocessor']
```

**Comment by andravin** [comment]
Thank you for tagging the related issue, that gave me a huge head start!

Maybe it is more pythonic for the CUDA function attributes to be properties of the object returned by RawKernel.attributes (or properties of RawKernel itself)?

The writeable attributes could have setters for the corresponding properties.

I followed this approach starting with the pull request from @grlee77 and now have something that works for me. Will submit a pull request later today so you can take a look.

**Comment by andravin** [comment]
Example usage:
```
op.fun.attributes.preferred_shared_memory_carveout = 75
op.fun.attributes.max_dynamic_shared_size_bytes = 48*1024

print('num_regs', op.fun.attributes.num_regs)
print('max_dynamic_shared_size_bytes', op.fun.attributes.max_dynamic_shared_size_bytes)
print('preferred_shared_memory_carveout', op.fun.attributes.preferred_shared_memory_carveout
```

**Comment by andravin** [comment]
The implementation is simpler still using descriptors:
```
class FunctionAttributes:

    class Read:
        def __init__(self, func_attribute):
            self.func_attribute = func_attribute

        def __get__(self, instance, owner):
            return driver.funcGetAttribute(self.func_attribute, instance.kern.ptr)

    class ReadWrite(Read):
        def __set__(self, instance, value):
            driver.funcSetAttribute(instance.kern.ptr, self.func_attribute, value)

    shared_size_bytes = Read(driver.CU_FUNC_ATTRIBUTE_SHARED_SIZE_BYTES)
    const_size_bytes = Read(driver.CU_FUNC_ATTRIBUTE_CONST_SIZE_BYTES)
    local_size_bytes = Read(driver.CU_FUNC_ATTRIBUTE_LOCAL_SIZE_BYTES)
    max_threads_per_block = Read(driver.CU_FUNC_ATTRIBUTE_MAX_THREADS_PER_BLOCK)
    num_regs = Read(driver.CU_FUNC_ATTRIBUTE_NUM_REGS)
    ptx_version = Read(driver.CU_FUNC_ATTRIBUTE_PTX_VERSION)
    binary_version = Read(driver.CU_FUNC_ATTRIBUTE_BINARY_VERSION)
    cache_mode_ca = Read(driver.CU_FUNC_ATTRIBUTE_CACHE_MODE_CA)
    max_dynamic_shared_size_bytes = ReadWrite(driver.CU_FUNC_ATTRIBUTE_MAX_DYNAMIC_SHARED_SIZE_BYTES)
    preferred_shared_memory_carveout = ReadWrite(driver.CU_FUNC_ATTRIBUTE_PREFERRED_SHARED_MEMORY_CARVEOUT)
    
    def __init__(self, kern):
        self.kern = kern
```

**Comment by leofang** [comment]
> Maybe it is more pythonic for the CUDA function attributes to be properties of the object returned by RawKernel.attributes (or properties of RawKernel itself)?

@andravin, It was request by a core dev to return a dict instead of attributes: https://github.com/cupy/cupy/pull/1874#discussion_r260107783

**Comment by andravin** [comment]
Well I thought I would just check in the code that works for me to see if it is useful to somebody else. You can do whatever you want with it.

One obvious advantage that properties have is that they can be set. If you return attributes as a dictionary, then you will also need a different interface for setting attributes. Makes sense just to use the standard python design pattern for getting and setting attributes instead.

Also, returning a dictionary is less efficient, because the whole dictionary is populated by calling `driver.funcGetAttribute` repeatedly, even if the user only wanted a single attribute. Using properties, you just pay for what you use.

While I am commenting on it, I don't think it makes sense for `cuda.device.attributes` to return a dictionary either.  That function dynamically populates the list of attributes, which is nice, but you could do that with properties and descriptors too, and it could be done at module import time instead of every time attributes is called for a new device. Basically a lot of extra work is being done to provide a less pythonic interface.

Anyway feel free to use any of my commits in any way they are useful. I might make some more changes in my fork to see how I like them. I suspect one way to clean up all this attributes stuff without breaking backwards compatibility is to add the "properties" directly to the `RawKernel` and `Device` classes. It makes literal sense that function-attributes should be properties of the `RawKernel` class, and device-attributes should be properties of the `Device` class.

Then the `attributes` property could be deprecated.

By the way, here is the best documentation I have found on python descriptors: https://docs.python.org/3/howto/descriptor.html

Anyway I hope this is useful.

**Comment by leofang** [comment]
Just to add a bit more background: without the proposed solution (#2120) we are unable to request full shared memory that newer devices can provide (in my case it's 96 KB=98304 bytes):
```python
In [38]: add_kernel((5,), (5,), (x1, x2, y, 25), shared_mem=98304)
---------------------------------------------------------------------------
CUDADriverError                           Traceback (most recent call last)
<ipython-input-38-a21f3e428b68> in <module>()
----> 1 add_kernel((5,), (5,), (x1, x2, y, 25), shared_mem=98304)

/home/leofang/test/cupy2/cupy/core/raw.pyx in cupy.core.raw.RawKernel.__call__()

/home/leofang/test/cupy2/cupy/cuda/function.pyx in cupy.cuda.function.Function.__call__()

/home/leofang/test/cupy2/cupy/cuda/function.pyx in cupy.cuda.function._launch()

/home/leofang/test/cupy2/cupy/cuda/driver.pyx in cupy.cuda.driver.launchKernel()

/home/leofang/test/cupy2/cupy/cuda/driver.pyx in cupy.cuda.driver.check_status()

CUDADriverError: CUDA_ERROR_INVALID_VALUE: invalid argument
```

**Comment by andravin** [comment]
Closing due to lack of interest.

**Comment by leofang** [comment]
This issue is addressed, see #2369.
